fmoe/linear.py

```diff
@@ -63,14 +63,14 @@ def __init__(
             self.bias = nn.Parameter(torch.zeros(num_expert, out_feat))
         else:
             self.register_parameter("bias", None)
-
+          
         self.reset_parameters()
 
     def forward(self, inp, fwd_expert_count):
         r"""
         Call MOE function
         """
-        x = MOELinear.apply(inp, fwd_expert_count, self.weight, self.bias)
+        x = MOELinear.apply(inp.float(), fwd_expert_count, self.weight, self.bias)
```

Comment:
I don't think we can convert `inp` to float. The weight may be half.

Reply:
I ran my code with a debugger and personally checked, I had the input as `float16` while weights and bias as `float32`.
I was pretty sure that AMP support needed CUDA modifications so a workaround would’ve been easier for now and with `float()` the code works fine, maybe `type_as` is more elegant so as soon as I can I’ll try it.

For sure one thing that can’t be done is to modify the weights and bias type as torch doesn’t like it when backpropagating.

Reply:
I think you should somehow make run `model.half()` for the `FMoELinear` module, instead of changing data types of weight and bias here.

Reply:
In fact I'm casting the input instead of the weights

Reply:
Yes that is what I was expecting, but the MoE layers are not casted, maybe the weights type are somehow hardcoded?
To say it better I'm uncasting the input back to float and in this way I'm able to train and it doesn't seem to have a big impact on the RAM

Reply:
The types are not hardcoded, but AMP may not correctly recognize it. I don't think we should change FastMoE like this to fit AMP.

Reply:
The alternative is to not have AMP compatibility, I think adding `inp.type_as(self.weight)` is just a small fix but it allows to have FastMoE working under any circumstances, in the end your code your choice.

Reply:
`inp.type_as(self.weight)` looks good to me.